File: JAVA/Print matrix in diagonal pattern.java
class Solution {
    public int[] matrixDiagonally(int[][] mat) {
        
        int n = mat.length;
        int ans[] = new int[n*n];
        int ind = 0;
        boolean up = true;
        
        int row = 0 , col = 0;
        
        while(row < n && col < n){
            
            if(up){
                while(row > 0 && col < n-1 ){
                    ans[ind++] = mat[row][col];
                    row--;
                    col++;
                }
                    ans[ind++] = mat[row][col];
                    if(col == n-1){
                        row++;
                    }
                    else{
                        col++;
                    }
                
            }else{
                
                while(row < n-1 && col > 0){
                    ans[ind++] = mat[row][col];
                    row++;
                    col--;
                    
                }
                    ans[ind++] = mat[row][col];
                    if(row == n-1){
                        col++;
                    }
                    else{
                        row++;
                    }
            }
            up = !up;
        }
        
        return ans;
        
    }
}

File: C++/Print matrix in diagonal pattern.cpp
class Solution{
  public:
    vector<int> matrixDiagonally(vector<vector<int>>&mat)
    {
        int r = 0;
        int c = 0;
        int n = mat.size();
        vector<int> ans;
        int count = 0;
        while(count < n*n){
            ans.push_back(mat[r][c]);
            count++;
            if(count >= n*n) break;
            if(c + 1 < n){
                c++;
            }
            else{
                r++;
            }
            while(r < n - 1 && c > 0){
                ans.push_back(mat[r][c]);
                count++;
                r++;
                c--;
            }
            ans.push_back(mat[r][c]);
            count++;
            if(r + 1 < n){
                r++;
            }
            else{
                c++;
            }
            while(r > 0 && c < n - 1){
                ans.push_back(mat[r][c]);
                count++;
                r--;
                c++;
            }
        }
        return ans;
    }
};
